app/Multitenancy/Tasks/ConfigurePasswordValidation.php

```diff
@@ -0,0 +1,138 @@
+                            $fail("The :attribute should contain at least {$settings->getMinDigits()} " . Str::plural('digit', $settings->getMinDigits()) . '.');
+                        }
+                    },
+                    function (string $attribute, mixed $value, Closure $fail) use ($settings) {
+                        if (preg_match_all('/[^A-Za-z0-9]/', $value) < $settings->getMinSpecialCharacters()) {
+                            $fail("The :attribute should contain at least {$settings->getMinSpecialCharacters()} special " . Str::plural('character', $settings->getMinSpecialCharacters()) . '.');
+                        }
+                    },
+                    function (string $attribute, mixed $value, Closure $fail) use ($settings) {
+                        $userToCheck = auth()->user();
+
+                        if (! ($userToCheck instanceof User)) {
+                            throw new Exception('User not found.');
+                        }
+
+                        if (filled($userToCheck->password) && Hash::check($value, $userToCheck->password)) {
+                            $fail('The new :attribute must be different from one you have used before.');
+
+                            return;
+                        }
+
+                        $numPreviousPasswords = $settings->getNumPreviousPasswords();
+
+                        if (! $numPreviousPasswords) {
+                            return;
+                        }
+
+                        $passwordHistory = $userToCheck->password_history ?? [];
+                        $passwordHistory = array_slice($passwordHistory, -$numPreviousPasswords);
+
+                        foreach ($passwordHistory as $oldPassword) {
+                            if (Hash::check($value, $oldPassword)) {
+                                $fail('The new :attribute must be different from one you have used before.');
+
+                                return;
+                            }
+                        }
+                    },
+                    function (string $attribute, mixed $value, Closure $fail) use ($settings) {
+                        if (! $settings->shouldBlacklistCommonPasswords()) {
+                            return;
+                        }
+
+                        $file = fopen(resource_path('text/common-passwords.txt'), 'r');
+
+                        if ($file === false) {
+                            throw new Exception('Could not open the file of common passwords.');
+                        }
+
+                        while (($commonPassword = fgets($file)) !== false) {
+                            if (Str::lower(trim($commonPassword)) === Str::lower(trim($value))) {
+                                fclose($file);
+
+                                $fail('The :attribute is too common. Please choose a different password.');
+
+                                return;
+                            }
+                        }
```

Comment:
Just to make this more readable, can we either:

A: Turn each of these closures to private functions in this class and reference them
B: Add a comment before each Closure naming/describing which rule it is for

Reply:
Done